### PR TITLE
sumeragi initial

### PR DESCRIPTION
**Pull request by eadventurous**
Signed-off-by: Egor Ivkov <e.o.ivkov@gmail.com>

### Description of the Change
Introduces initial structure of Sumeragi module and also implements a base case of consensus, when everything goes according to plan.

### CODE REVIEW

File: iroha/src/crypto.rs [hunk 1]
```diff
@@ -5,6 +5,25 @@ pub type Hash = [u8; 32];
 pub type PublicKey = [u8; 32];
 type Ed25519Signature = [u8; 64];
 
+pub struct Crypto;
+
+impl Crypto {
+    pub fn hash(bytes: Vec<u8>) -> Hash {
+        use ursa::blake2::{
```
Comment (humb1t) [inline iroha/src/crypto.rs:12]: Please, put use on top of the file.

File: iroha/src/lib.rs [hunk 2]
```diff
@@ -60,7 +62,16 @@ impl Iroha {
             Arc::clone(&world_state_view),
             transactions_sender,
         );
-        let sumeragi = Sumeragi::new();
+        //TODO: get peers from blockchain
```
Comment (humb1t) [inline iroha/src/lib.rs:65]: As I remember, @takemiyamakoto assumed we will have additional JSON file with trusted peers.

File: iroha/src/lib.rs [hunk 3]
```diff
@@ -112,7 +123,7 @@ impl Iroha {
                             sumeragi
                                 .lock()
                                 .await
-                                .sign(transactions)
+                                .sign_transactions(transactions)
```
Comment (humb1t) [inline iroha/src/lib.rs:126]: if we have only one function which starts with `sign` we can stay with this naming. because right now it reads as - sign_transactions_transactions.

File: iroha/src/peer.rs [hunk 4]
```diff
@@ -56,9 +56,10 @@ impl Ping {
     }
 }
 
-#[derive(Encode, Decode, PartialEq, Eq, Debug, Clone, Hash)]
-struct PeerId {
-    listen_address: String,
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Clone, Hash, Io)]
+pub struct PeerId {
+    pub listen_address: String,
```
Comment (humb1t) [inline iroha/src/peer.rs:61]: maybe just an address?

File: iroha/src/sumeragi.rs [hunk 5]
```diff
@@ -1,14 +1,214 @@
...
-    pub fn new() -> Self {
-        Sumeragi {}
+    pub fn new(
+        peers: &[PeerId],
+        prev_block_hash: Option<Hash>,
+        max_faults: usize,
+    ) -> Result<Self, String> {
+        if peers.len() >= 3 * max_faults + 1 {
+            let mut peers = peers.to_vec();
+            Self::sort_peers(&mut peers, prev_block_hash);
```
Comment (humb1t) [inline iroha/src/sumeragi.rs:44]: Maybe we can use sorted collection?

File: iroha/src/lib.rs (hunk 3 above)
Comment (eadventurous) [reply iroha/src/lib.rs:126]: There is also a function `sign_block` in Sumeragi.

File: iroha/src/sumeragi.rs (hunk 5 above)
Comment (eadventurous) [reply iroha/src/sumeragi.rs:44]: Ordering is determined not only by PeerId but also by previous block hash, so we still would need to resort it every new round. Taking into account that we sort all the peer only in the beginning of each round, I don't think using sorted collection would improve anything.

Comment (humb1t) [reply iroha/src/sumeragi.rs:45]: https://soramitsu.atlassian.net/browse/RH2-67

File: iroha/src/block.rs [hunk 6]
```diff
@@ -33,19 +35,7 @@ impl Block {
...
-
-        let bytes: Vec<u8> = self.into();
-        let vec_hash = VarBlake2b::new(32)
-            .expect("Failed to initialize variable size hash")
-            .chain(bytes)
-            .vec_result();
-        let mut hash = [0; 32];
-        hash.copy_from_slice(&vec_hash);
-        hash
+        Crypto::hash(self.into())
```
Comment (takemiyamakoto) [inline iroha/src/block.rs:38]: To keep PRs a bit smaller, I think next time you can put crypto refactor into a separate PR from consensus logic.

File: iroha/src/sumeragi.rs [hunk 7]
```diff
@@ -1,14 +1,216 @@
...
+pub struct Sumeragi {
+    sorted_peers: Vec<PeerId>,
+    max_faults: usize,
+    peer_id: PeerId,
+    /// Block in discussion this round
+    pending_block: Option<Block>,
+}
+
+#[derive(Io, Decode, Encode, Debug, Clone)]
+pub enum Message {
```
Comment (takemiyamakoto) [inline iroha/src/sumeragi.rs:19]: For error handling, you also need to add:
- CreateLeaderSuspect/SignLeaderSuspect
- CreateCommitTimout/SignCommitTimeout
- CreateVotingTimeout/SignVotingTimeout

Here is some info about each:
LeaderSuspect
- f+1 signatures are needed to elect a new leader
- used when someone sends a tx to the leader but a new block is not created within some expected time (like the block time)

CommitTimout
- f+1 signatures are needed and if f+1 are received on this message, the f+1 signatures and the block hash are written into the successfully created block; this is done to prevent forking of stake via a committed block withholding attack
- every time a peer votes for a block, they expect a commit message within a certain amount of time and this timeout occurs if they don't get a commit message when expected; the solution is therefore to invalidate their vote and the old block and elect a new leader and proxy tail (probably randomly recording everyone so that a new leader and new proxy tail are elected is good. obviously any random reordering needs to be deterministic as every peer will need to compute the new order independently)

VotingTimeout
- this message can only be created by the proxy tail
- the proxy tail starts a timer when they receive first vote or block proposal and if the 2f+1 votes for a block are not all received before the timer goes off, the proxy tail creates this voting timeout and sends it to all the peers
- we MAY be able to get rid of this as a separate type as the commit timeout should handle this as well; let's explore this simplification

File: iroha/src/block.rs (hunk 6 above)
Comment (eadventurous) [reply iroha/src/block.rs:38]: Thanks, will take into account next time.

File: iroha/src/sumeragi.rs (hunk 7 above)
Comment (eadventurous) [reply iroha/src/sumeragi.rs:19]: Added this information to https://wiki.hyperledger.org/display/iroha/Sumeragi+Error+Handling

File: iroha/src/sumeragi.rs [hunk 8]
```diff
@@ -1,14 +1,216 @@
...
+    peer::{self, PeerId},
+    prelude::*,
+};
+use iroha_derive::*;
+use parity_scale_codec::{Decode, Encode};
+use std::cmp::Ordering;
 
-#[derive(Default)]
-pub struct Sumeragi {}
+pub struct Sumeragi {
```
Comment (takemiyamakoto) [inline iroha/src/sumeragi.rs:10]: Probably need to add state info about how many failures there were in the current round (index) to know the current order of the nodes after a leader election.

File: iroha/src/sumeragi.rs [hunk 9]
```diff
@@ -1,14 +1,216 @@
...
+            .first()
+            .expect("Failed to get first peer.")
+    }
+
+    pub fn proxy_tail(&self) -> &PeerId {
+        self.peers_set_a().last().expect("Failed to get last peer.")
+    }
+
+    pub fn validating_peers(&self) -> &[PeerId] {
+        &self.peers_set_a()[1..(self.peers_set_a().len() - 1)]
```
Comment (takemiyamakoto) [inline iroha/src/sumeragi.rs:86]: what happens if the peers_set_a is Len 0?

File: iroha/src/sumeragi.rs [hunk 10]
```diff
@@ -1,14 +1,216 @@
...
+            Ok(())
+        } else {
+            Err(format!(
+                "Peer needs to be one of {:?} for this operation.",
+                allowed_roles
+            ))
+        }
+    }
+
+    fn sort_peers(peers: &mut Vec<PeerId>, block_hash: Option<Hash>) {
```
Comment (takemiyamakoto) [inline iroha/src/sumeragi.rs:112]: It might be more efficient to seed a random number generator with the previous block hash and then use that to determine the order.

File: iroha/src/sumeragi.rs (hunk 9 above)
Comment (eadventurous) [reply iroha/src/sumeragi.rs:86]: This can't happen as the `Sumeragi::new` function checks that the number of peers is more or equal to `3f+1`, so it's at least 1.
